go/tools/builders/cover_test.go

```diff
@@ -113,3 +113,44 @@ func TestRegisterCoverage(t *testing.T) {
 		}
 	}
 }
+
+func TestFindFirstLineDirectiveFilename(t *testing.T) {
+	tests := []struct {
+		desc string
+		src string
+		content string
+		want string
+	}{
+		{desc: "no line directive", src: "noline.go", content: "package main\n\nfunc main() {}\n", want: ""},
+		{desc: "slash slash directive", src: "slashslash.go", content: "package main\n\n//line foo.go:123\n\nfunc main() {}\n", want: "foo.go"},
+		{desc: "slash star directive", src: "slashstar.go", content: "package main\n\n/*line foo.go:123*/", want: "foo.go"},
+		{desc: "windows absolute path", src: "windows.go", content: "package main\n\n//line C:\\Windows\\path.go:300", want: "C:\\Windows\\path.go"},
+		{desc: "windows with column", src: "windows.go", content: "package main\n\n//line C:\\Windows\\path.go:300:10", want: "C:\\Windows\\path.go"},
+		{desc: "url-like path", src: "url.go", content: "package main\n\n//line http://example.com/file.go:400", want: "http://example.com/file.go"},
+		{desc: "multiple colons", src: "multiple.go", content: "package main\n\n//line scheme:rest:file.go:500", want: "scheme:rest:file.go"},
+		{desc: "empty filename", src: "empty.go", content: "package main\n\n//line :50", want: ""},
+		{desc: "invalid line number", src: "invalid.go", content: "package main\n\n//line file.go:0", want: ""},
+		{desc: "invalid line number", src: "invalid.go", content: "package main\n\n//line file.go:invalid", want: ""},
+		{desc: "not a directive", src: "notadirective.go", content: "package main\n\nnot a line directive", want: ""},
+		{desc: "incomplete", src: "incomplete.go", content: "package main\n\n//line", want: ""},
+		{desc: "missing line number", src: "missing.go", content: "package main\n\n//line file.go:", want: ""},
+	}
+
+	for _, test := range tests {
+		testTempDir := t.TempDir()
+		testFile := filepath.Join(testTempDir, test.src)
+		err := os.WriteFile(testFile, []byte(test.content), 0644)
+		if err != nil {
+			t.Errorf("create test file: %v", err)
+		}
+		defer os.Remove(testFile) // Clean up the test file.
+
+		got, err := findFirstLineDirectiveFilename(testFile)
+		if err != nil {
+			t.Errorf("find first line directive filename: %v", err)
+		}
+		if got != test.want {
+			t.Errorf("find first line directive filename: got %q, want %q", got, test.want)
+		}
+	}
+}
```

Comment:
Nit: I don't think testing an unexported function like this is useful. It forces future maintainers/contributors to study the implementation details when doing refactoring.

I would prefer it if we could add a high-level test, say in `tests/core/coverage/`, that verifies the desired behavior. The setup could be as simple as running coverage over a `go_test` target with `genrule` created one of the source files. That way, we can check for regression in future releases.